### ISSUE TITLE
GraphBLAS JIT isn't working in MINGW when using cmake

### ISSUE THREAD

**Issue by DrTimothyAldenDavis** [DrTimothyAldenDavis/SuiteSparse#572]
The GraphBLAS JIT is encountering compiler errors when compiling the JIT kernels on MINGW when using cmake.   The diagnostics are pretty terse but this occurs when GraphBLAS tries to use cmake to build a kernel.  The name configured by cmake to create GraphBLAS/Config/GB_config.h must be inserting this line:

`#define GB_C_COMPILER " /clang64/bin/cc.exe"
`

The failed output is below.  The gauss_demo still succeeds, because of how I handle compiler errors in the JIT.  If a JIT kernel fails to compile or load, it's hard to tell inside GraphBLAS exactly what the problem is.  It might be a syntax error in the user-defined operator, for example.  Since I have a function pointer anyway, I can fallback to using a "generic" kernel that uses this function pointer, and then I return the correct result to the user application.  I turn off the JIT in that case to avoid repeated errors.  So everything still works, just slower with no JIT.

I should revise my tests to assert that after compiling a valid kernel (like this one), that the JIT is still enabled.  Then I would have caught this error as a test failure.

The GB_C_COMPILER string is set by GraphBLAS/cmake_modules/GraphBLAS_JIT_configure.cmake, and there's a special case there for MINGW.

It's possible that the GB_COMPILER string is OK but this line is causing the problem for MINGW:

https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/0b41a3e842271e1b180447f42a185f3469e7d21d/GraphBLAS/Source/GB_jitifyer.c#L2268-L2276

Here's the log with the error.  I will add some more diagnostics to try to track this down (I can print out all my calls to `system(...)` for example:

```
  CMake Error at CMakeLists.txt:2 (project):
    The CMAKE_C_COMPILER:
  
      /clang64/bin/cc.exe
  
    is not a full path to an existing compiler tool.
  
    Tell CMake where to find the compiler by setting either the environment
    variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
    the compiler, or to the compiler name if it is in the PATH.
  
  
  ninja: error: loading 'build.ninja': The system cannot find the file specified.
  
  
  CMake Error: Error processing file: C:/msys64/home/runneradmin/.SuiteSparse/GrB8.3.0/tmp/ce2be202506bf1b4/cmake_install.cmake
  gauss_demo: all tests pass
```

**Comment by DrTimothyAldenDavis** [comment]
Here are more diagnostics:

```
system: cmake -S "C:/msys64/home/runneradmin/.SuiteSparse/GrB8.3.0/tmp/ce2be202506bf1b4" -B "C:/msys64/home/runneradmin/.SuiteSparse/GrB8.3.0/tmp/ce2be202506bf1b4" -DCMAKE_C_COMPILER="/clang64/bin/cc.exe"   
  CMake Error at CMakeLists.txt:2 (project):
    The CMAKE_C_COMPILER:
  
      /clang64/bin/cc.exe
  
    is not a full path to an existing compiler tool.
  
    Tell CMake where to find the compiler by setting either the environment
    variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
    the compiler, or to the compiler name if it is in the PATH.
  
  
  result: 1
  system: cmake --build "C:/msys64/home/runneradmin/.SuiteSparse/GrB8.3.0/tmp/ce2be202506bf1b4" --config Release   
  ninja: error: loading 'build.ninja': The system cannot find the file specified.
  
  
  result: 1
  system: cmake --install "C:/msys64/home/runneradmin/.SuiteSparse/GrB8.3.0/tmp/ce2be202506bf1b4"   
  CMake Error: Error processing file: C:/msys64/home/runneradmin/.SuiteSparse/GrB8.3.0/tmp/ce2be202506bf1b4/cmake_install.cmake
  result: 1
```

It looks like the compiler name "/clang64/bin/cc.exe" works when using direct compilation, but not as an input to cmake with 

`cmake -DCMAKE_C_COMPILER="/clang64/bin/cc.exe" ...
`

For the latter, it might be that the compiler string needs to be the Windows variant path, not the POSIX path, so this conversion is breaking things:

https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/0b41a3e842271e1b180447f42a185f3469e7d21d/GraphBLAS/cmake_modules/GraphBLAS_JIT_configure.cmake#L10-L17

I may need to provide a way to specify two compiler strings:  one to use for cmake, and one to use for a direct compile.

I'd like to get the MINGW system to work with both ways of compiling a JIT kernel: via cmake and via a direct compilation.  @mmuetzel : can you let me know if this fix would work for MINGW?  It's a fair bit of work to add another string to GraphBLAS; I would need to add a GrB_set / GrB_get option, an enum for it , tests, and so on.  That shouldn't be too hard but I want to ask you what you think before I try adding it.

I could actually test it by hacking the gauss_demo program, and checking the result in the MING CI on github.  Whenever I use cmake for a JIT kernel, I could use GrB_set to set the C compiler to a hard coded string:

`"C:/msys64/clang64/bin/cc.exe"`

and then set it back to

`"/clang64/bin/cc.exe"`

when using a direct compilation.

**Comment by mmuetzel** [comment]
Using the full Windows-style path to the compiler would cause relocation issues (at least for the version of GraphBLAS distributed by MSYS2).
Alternatively, it might help to call the `cmake` command from a (POSIX) shell. I.e., add `sh -c` at the front of each `cmake` command (at least for MinGW). The POSIX shell of MSYS2 would "recognize" the POSIX style path and should be able to find the corresponding executable.

**Comment by DrTimothyAldenDavis** [comment]
Thanks for the suggestion-- I'll give that a try.

**Comment by mmuetzel** [comment]
Another question: Did you force it to use CMake in MinGW? At least at some point, it was working without CMake for me. Shouldn't that be preferred? Or is the direct compilation also broken for MinGW?

**Comment by DrTimothyAldenDavis** [comment]
Yes, the demo forces the use of cmake. The direct compile is always preferred and is the default on all but MSVC. 

The gauss_demo exercises both, but mainly forces the use of cmake.  Just as a test.

**Comment by DrTimothyAldenDavis** [comment]
Direct compile works fine in MINGW. The wild_type demo uses the default exclusively.

**Comment by mmuetzel** [comment]
I went ahead and opened #574 with a change that runs `cmake` through `sh` on MinGW for proper path translation.

**Comment by DrTimothyAldenDavis** [comment]
I'll try to get it to work.

The main reason I added the cmake option for building JIT kernels is that I couldn't get any direct compilation working on Windows (using MSVC).  So I added the cmake method, and then enabled it as an option for Linux/Mac in case there was any issues on those platforms.  The direct compilation option could be more temperamental, I thought, because cmake is more portable and more uniform.

So I thought ...

For GraphBLAS 8.3.0 and SuiteSparse 7.4.0, we do have a method on MINGW that works for JIT kernels, which is the direct compile.  That's faster than cmake anyway.  

Well, one option at least for this release is to allow MINGW to *only* compile with direct compilation and *not* allow the use of cmake.  If GraphBLAS is on MINGW and a request is made to compile using cmake, I could just ignore it.  I do the opposite for MSVC: using cmake is the only option.  I could just make the reverse happen for MINGW.

Then the gauss_demo would work, just like it already does for the wildtype_demo, which uses the defaults (direct compile on Linux, Mac, and MINGW, and cmake on MSVC).

Then I could add a future TODO to enable cmake for MSVC.

**Comment by DrTimothyAldenDavis** [comment]
Looking over the log file, I think I will punt, and simply disable the use of cmake to build JIT kernels on MINGW. 

So for now:
MINGW:  direct compile only
MSVC: cmake only
Linux and Mac: either one

I can figure out later how to get MINGW to use cmake, and MSVC to use a direct compile, for compiling the JIT kernels.  That way this issue won't hold up a SuiteSparse 7.4.0.beta1 release.

**Comment by mmuetzel** [comment]
That sounds like a good-enough solution to me.

For later reference, the output in the CI was:
```
  starting gauss_demo
  system: sh -c cmake -E remove_directory "C:/msys64/home/runneradmin/.SuiteSparse/GrB8.3.0/tmp/ce2be202506bf1b4"   
  result: 0
  system: sh -c cmake -S "C:/msys64/home/runneradmin/.SuiteSparse/GrB8.3.0/tmp/ce2be202506bf1b4" -B "C:/msys64/home/runneradmin/.SuiteSparse/GrB8.3.0/tmp/ce2be202506bf1b4" -DCMAKE_C_COMPILER="/mingw64/bin/cc.exe"   
  result: 0
  system: sh -c cmake --build "C:/msys64/home/runneradmin/.SuiteSparse/GrB8.3.0/tmp/ce2be202506bf1b4" --config Release   
  result: 0
  system: sh -c cmake --install "C:/msys64/home/runneradmin/.SuiteSparse/GrB8.3.0/tmp/ce2be202506bf1b4"   
  result: 0
  system: sh -c cmake -E remove_directory "C:/msys64/home/runneradmin/.SuiteSparse/GrB8.3.0/tmp/ce2be202506bf1b4"   
  result: 0
  info: -3 error! File D:/a/SuiteSparse/SuiteSparse/GraphBLAS/Demo/Program/gauss_demo.c, Line 274
```

I'd guess that `result: 0` means that each of the cmake calls succeeded.
I don't know what `info: -3 error!` means.

**Comment by DrTimothyAldenDavis** [comment]
The -3 was actually an glitch.  I was using a result "info" that was uninitialized, by mistake.  It should have been from the prior call to GraphBLAS.

So the value -3 is bogus, but the main issue here is that the JIT control was changed.  I change the JIT from "on" to "off" if I encounter any kind of error when compiling OR loading a *.so OR looking up the JIT function in the just-loaded *.so (or dylib or dll).

So in this case, something broke.  It didn't break in the call to system("... string...").  It broke later.  That means something in the build itself broke, but it's not clear to me what broke.  I'd need to look at the detailed output.

GraphBLAS doesn't try to figure out why something breaks.  There are just too many ways the JIT process can break.  I just do:

(1) write the file.c to the cache folder
(2) create the cmake script (for this case), also in the cache folder
(3) use system ('sh -c cmake ...') to compile the library
(4) dlopen the library on Linux/Mac, or LoadLibrary on Windows
(5) dlsym to get the function pointer on Linux/Mac, or GetProcAddress on Windows

If the JIT function pointer is NULL:  I don't have a kernel.  Oh well, turn off the JIT (so I don't get repeated failures), and just punt to a generic kernel.   The generic kernel is built in, but instead of using a compiled JIT kernel (one call through a function pointer), it uses function pointers for each operation (like a single add, a single multiply, ...).  That works but is slow.

It's too hard to provide a detailed analysis of what broke, and I have a fallback if I get a NULL at the end of all that, anyway.

I turn off the JIT because the failure can be systemic, like "I don't have a valid string that gives me the name or path of the compiler", or "cmake is not installed" etc.

I would have to dig deeper here to find out why I can't use cmake with MINGW.  It would take a careful look at each step, and it's not something GraphBLAS itself does.  It's hard for me to do since my access to MINGW is all through the github CI.  And it's not super critical, since direct compile works and is much faster than cmake anyway.

It would be more critical to figure out how to get a direct compile to work in MSVC.  I don't use Windows much so that's a challenge for me.

**Comment by DrTimothyAldenDavis** [comment]
This is fixed (in a way) in the current GraphBLAS 8.3.0 and SuiteSparse 7.4.0.beta1, simply by doing all JIT compilation using direct compilation (which is better anyway).  I have a TODO tagged in the code; it would be nice to get cmake working in the future, but it's not so critical.
